### DISCUSSION TITLE
Sudden Latency

### DISCUSSION THREAD

**Discussion by jamesonuk** [Koenkk/zigbee2mqtt#24214] [General]
I have been using Z2M for several years and it has been rock solid but I did the cardinal sin of changing more than one thing at once yesterday and now I am experiencing lag (usually ~3s but anything up to ~30s+).   Previously the same actions were essentially instant.

Yesterday I did a few things:

- Updated Z2M to 1.40.2 (from 1.40.1)
- Updated HA to 2024.10.0 (from 2024.9.3
- Updated Mosquitto to 2.0.19 (from 2.0.18)
- Applied the latest updates to Ubuntu Mantic Minotaur
- Added a Sonoff SNZB-04P contact sensor to my mesh

Everything is running on Docker on Ubuntu and using a Conbee II stick

It was after doing all the updates I started noticing the things weren't working.   Things I have bound at Zigbee level work instantly but if I try and control anything from HA it doesn't work or takes an age to respond.

I am seeing lots of the following in Z2M logs
```
[2024-10-03 15:19:29] error:    z2m: MQTT error: Keepalive timeout
[2024-10-03 15:19:30] error:    z2m: Not connected to MQTT server!
[2024-10-03 15:19:30] error:    z2m: Cannot send message: topic: 'zigbee2mqtt/Occupancy Sensor', payload: '{"detection_delay":0,"fading_time":2,"illuminance_lux":18,"last_seen":"2024-10-03T15:19:30.120Z","linkquality":255,"maximum_range":0.75,"minimum_range":0,"presence":false,"radar_sensitivity":7,"target_distance":0}
[2024-10-03 15:19:30] info:     z2m: Connected to MQTT server
```
and much less frequently
```
14128494-[2024-10-04 11:54:41] error:   z2m: Not connected to MQTT server!
14128569:[2024-10-04 11:54:43] error:   z2m: MQTT error: connack timeout
14128642-[2024-10-04 11:54:44] info:    z2m: Connected to MQTT server
```
which suggests and MQTT issue but I then had a play to try and control devices via Z2M UI and this seemed better but still kept seeing period of unresponsiveness.

So I rolled back all the containers to the previous versions (at first one at a time but then in desperation all of them) but the issue is still there.    I have ended up powering down the entire house to reset everything but still it is there.

I have removed the Sonoff device I added, restarted umpteen times and generally run out of ides of where to look.

Any advice on where to look at debugging this?

**Comment by chris-1243** [comment]
Would you mind sharing the `mqtt` section of `configuration.yaml` ?

Also, check the mqtt log to see what is going on there. I am also running everything in Docker and latest version without issues

**Reply by jamesonuk** [reply]
> Would you mind sharing the `mqtt` section of `configuration.yaml` ?

```yaml
mqtt:
  base_topic: zigbee2mqtt
  server: mqtts://mosquitto:8883
  ca: /app/data/root-ca.pem
  key: /app/data/ha_mqtt.key.pem
  cert: /app/data/ha_mqtt.cert.pem
  keepalive: 60
  reject_unauthorized: true
  version: 4
```
nothing has changed config wise and if it was version related then flashing back versions should have resolved it...

I don't think it is necessarily related to version (again as rolling everything back made no difference) I am just stumped as to where to go.

Mosquitto logs have
```
2024-10-04T20:39:51.652755213Z 1728074391: OpenSSL Error[0]: error:0A000126:SSL routines::unexpected eof while reading
2024-10-04T20:39:51.652802168Z 1728074391: Client mqttjs_72daf128 closed its connection.
2024-10-04T20:39:52.662130717Z 1728074392: New connection from 10.99.99.4:49090 on port 8883.
2024-10-04T20:39:52.674490487Z 1728074392: New client connected from 10.99.99.4:49090 as mqttjs_72daf128 (p2, c1, k60, u'ha_mqtt').
```
but I think that is down to the connection crashing more than anything else.

The other thing that stands out in the logs is that whenever this happens, there is almost always a connect message the same second

**Reply by chris-1243** [reply]
> OpenSSL Error[0]: error:0A000126:SSL routines::unexpected eof while reading

Your client did not close its connection correctly. You definitely have a communication problem between Z2M and your MQTT broker.

Are you using self-signed certificate?
If yes, I would check this in your `mqtt` section according to the [documentation](https://www.zigbee2mqtt.io/guide/configuration/mqtt.html#server-connection)

```
# Optional: disable self-signed SSL certificates (default: true)
    reject_unauthorized: true
```

It should be `false` if my understanding is correct.

As you are on Docker, maybe try to launch a new mqtt container with any user/password or ssl to check if everything works correctly. Of course, a full backup before doing this has to be done😉

**Reply by lux73** [reply]
if all your Stuff is inside your LAN there is no need to use Certificates - kick them out

**Reply by jamesonuk** [reply]
> Your client did not close its connection correctly. You definitely have a communication problem between Z2M and your MQTT broker.

Exactly but Z2M shouldn't be closing connections so it is a symptom not the actual issue  (Thinking about it these could just be a distraction and they could just tie in with when I restarted container.    Either way I don't see these are any help.
 
> Are you using self-signed certificate?

Yes but I have a CA so they are trusted and as I said this has been working for several years without issue
`ca: /app/data/root-ca.pem` sorts out the trust issue

**Reply by jamesonuk** [reply]
> if all your Stuff is inside your LAN there is no need to use Certificates - kick them out

Way OT and perhaps due to the fact I do a lot of work with Cyber Security folk but there are good reasons to apply trust on your local LAN.   Probably less important for MQTT but it is not much work and provides additional levels of security and privacy

**Reply by lux73** [reply]
I basically agree with you, but if someone has already made it into your LAN - you have completely different problems than your mqtt ssl encryption

mqtt ssl inside LAN just complicates things without really generating any security - using User and Passwords is already enough

...but thats only my Side of view 😉

**Reply by jamesonuk** [reply]
> I basically agree with you, but if someone has already made it into your LAN - you have completely different problems than your mqtt ssl encryption

Indeed, https is a far more important.   If someone has made it on to your LAN (and lets face it with automation we probably have lots of devices that are susceptible to being taken over and Matter can only make this worse....)  then using http means that everything is flying around unencrypted and it is far easy to start sniffing username / passwords.   MQTTT is probably the lowest risk thing but still using a certificate makes it a bit harder for anyone on my LAN flicking my smart plugs.     Still I do middleware and platform engineering so I am just doing things that I would expect in an enterprise environment (VLANs, zero trust, monitoring ....)    It really isn't that much additional work

**Comment by jamesonuk** [comment]
I do seem to have established that it is HA that is causing my MQTT issues (which is odd)
I did rollback the other day but it didn't seem to work but I think this might have been a combination of things.

I was messing around last night and found that starting everything up bar HA and everything was stable.   Controlling things via Z2M worked instantly and was stable for at least 30 minutes.    I then started up HA and even controlling via Z2M had lag and started seeing disconnects.  Oddly though I don't think that was the actual issue as even controlling directly via Z2M was experiencing lag.   (I am presuming that if you turn a light on via Z2M UI then it will send the Zigbee message before sending a MQTT message)

I am struggling to comprehend what the underlying cause of this could be but it is pointing towards something OS related being triggered by HA

**Reply by jamesonuk** [reply]
Really odd as trying to break things again I have re-installed HS 2024.10 and everything is still working.....

No amount of restarting (containers or whole box) was helping yesterday but seems fine today.

Perhaps I shouldn't do updates on a Friday again....

Really weird as I just cannot see what could have caused things to behave like this
